### DISCUSSION TITLE
[General] Failed transaction on ERC20 token transfer with paymaster.

### DISCUSSION THREAD

**Discussion by mshojaei-txfusion** [zkSync-Community-Hub/zksync-developers#253] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

We deployed an approval-based paymaster, and since then it managed to handle more than 800 transactions, but one specific user when tried to transfer some ZKUSD tokens, the transaction failed, we also tried to reproduce the error but we successfully managed to do the same transaction.

### What is the specific issue or error you're encountering?

- ERC20 Token is transferred to the paymaster contract
- Transaction was erc20 token transfer of [ZKUSD](https://explorer.zksync.io/address/0xfC7E56298657B002b3e656400E746b7212912757#transactions)
- Transaction is failed with this error `Error function_selector = 0x, data = 0x`

### Can you share the error messages or logs you're receiving, if any?

This is the transaction address: https://explorer.zksync.io/tx/0x37eb2a44bd3c71a5a42d3a89adc9500dbda4e2a512e0286c9fdec3455d1f469c

Status: `Failed`
Reason: `Error function_selector = 0x, data = 0x`

### Have you made any recent changes to the contract before encountering this issue?

Not at all.
all clients and smart contracts are the same as before (when the other 800 transactions with Paymaster succeed).

### Are there any external libraries or contracts that your contract interacts with?

The user tried to transfer the ERC20 token and also used Paymaster.
ERC20 Token Contract: https://explorer.zksync.io/address/0xfC7E56298657B002b3e656400E746b7212912757
Paymaster address: https://explorer.zksync.io/address/0x5F0B9f4f6BaB83521170A9C82610A9F006625B6F

### Can you provide the relevant portions of your contract code where the issue is occurring?

we believe that the problem is not related to the paymaster, but also do not have any idea why this happened.
anyway our paymaster is similar to this paymaster example (but not fixed price, connected to a price feed): https://github.com/matter-labs/paymaster-examples/blob/main/contracts/contracts/paymasters/ERC20fixedPaymaster.solhappened

### Have you tried to isolate the problem, and if so, what were the results?

No, could not even reproduce the problem, but successfully managed to transfer ZKUSD with the same paymaster: 
https://explorer.zksync.io/tx/0xd45c13a309aac609052ea30a93a300e9b771baa55aa7cf82934237e07e9621b0

### What steps have you already taken to try to resolve the issue?

Before there was a similar problem that randomly failing transactions with Paymaster, the ML team said the reason behind it was sometimes transactions did have not enough gas fees that other transactions could cover their lack of gas fees, so they said the problem was with estimation, and then the problem was fixed by topping up the estimations.
but in this case, if you check the transaction link I sent you can see that a good amount of the transaction fee is refunded.
so the problem can not be having enough transaction fees.

![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/138107084/b2992877-f0f5-4b65-ad8e-996b92757e96)

### Repo Link (Optional)

portal.txsync.io

**Comment by ZKBitcoinToken** [comment]
I have been submitting tons of answers to my paymaster and I get failed transactions all the time on it.  It happens when two concurrent answers are submitted that are both valid at the time, but once one is accepted the other becomes invalid.  This invalid transaction will still go to the paymaster and FAIL and make the paymaster pay fee.  I think this is a big error

**Reply by ZKBitcoinToken** [reply]
https://goerli.explorer.zksync.io/tx/0xa8b62ec56ae1346fe93ed49d2ba0437c5eb1a491c9473d0d29ac5bdba08ddf89
https://goerli.explorer.zksync.io/tx/0xa9715d333b7075ab7cb5aa6f493ce709e92f5ecbfd4aaf917b1b750349a30911
https://goerli.explorer.zksync.io/tx/0xb2333ea661f7376086a6f145adfef3b7317ff662cb32b8d47ec252e14a8d18c2
https://goerli.explorer.zksync.io/tx/0xa9715d333b7075ab7cb5aa6f493ce709e92f5ecbfd4aaf917b1b750349a30911

Failed transactions PAID BY PayMaster
I can recreate easily

**Reply by MexicanAce** [reply]
> It happens when two concurrent answers are submitted that are both valid at the time, but once one is accepted the other becomes invalid [...]

@ZKBitcoinToken can you expand what you mean by "two concurrent answers"? If you have steps for recreating this issue, would you please post them in this thread for us to dig in further?

**Reply by ZKBitcoinToken** [reply]
Sorry was wrong about this, my PayMaster wasnt checking things correctly,

**Comment by MexicanAce** [comment, marked as answer]
> anyway our paymaster is similar to this paymaster example

@mshojaei-txfusion Could you please post the code for the paymaster so we can take a closer look? The error `Error function_selector = 0x, data = 0x` can be generic at times, but usually refers to accessing methods that don't exist.

**Reply by MexicanAce** [reply]
It's also possible to investigate further using the `era_test_node` tool to replay the transaction locally for further details: https://era.zksync.io/docs/tools/testing/era-test-node.html#replay-remote-transactions-locally

**Comment by ZKBitcoinToken** [comment]
> > It happens when two concurrent answers are submitted that are both valid at the time, but once one is accepted the other becomes invalid [...]
> 
> @ZKBitcoinToken can you expand what you mean by "two concurrent answers"? If you have steps for recreating this issue, would you please post them in this thread for us to dig in further?
I actually was mistaken, I think we need to add a

 I think I am mistaken and we need a 
 
        require(gasleft() > GAS_BUFFER2, "Not enough gas left to safely proceed, please raise your gasLimit before sending");
But his error doesnt seem related to out of gas? So I am not sure, it doesnt seem like there was any error in the transfer (the user had that amount of USDC, maybe the transfer attempted to move too much UDSC or other asset he didnt have causing it to error??

**Comment by teschdV** [comment]
# It seems like the transaction failed with the error "Error function_selector = 0x, data = 0x". This could indicate an issue with the data or function selector provided in the transaction. 
# Since you've mentioned that you couldn't reproduce the problem and the Paymaster has been handling transactions successfully before, it's possible that this specific transaction had some anomaly in its data.
# One possibility is that the user might have provided incorrect data or function selector when initiating the transaction. It's also worth checking if there were any changes in the ERC20 token contract that could have affected this transaction.
# Further investigation into the specific details of the failed transaction, such as the data payload and function selector, could provide insights into resolving this issue.

**Comment by powstingding** [comment]
Investigating a specific failed transaction with an approval-based paymaster for ERC20 token transfer. Despite similar successful transactions, this one resulted in a "function_selector = 0x, data = 0x" error. Exploring gas estimation, contract interactions, and potential edge cases may help identify the root cause.

**Comment by tashuraschaefer** [comment]
It seems that the transaction failed due to an empty function selector and data. This could be caused by various factors, such as incorrect input parameters or unexpected behavior in the contract. Further investigation and debugging are needed to determine the root cause of the issue.

**Comment by msandiferv** [comment]
The specific transaction failed with the error "function_selector = 0x, data = 0x" on zkSync Mainnet. 
Despite successful previous transactions, this particular one encountered an issue. 
Investigation suggests that sufficient gas fees were provided, ruling out that as the cause.
Further analysis and consultation with the zkSync team may be required to identify the root cause.
